packages/jitm/tests/php/test_JITM.php

```diff
@@ -0,0 +1,204 @@
+<?php
+
+namespace Automattic\Jetpack;
+
+use Automattic\Jetpack\JITM;
+use phpmock\functions\FunctionProvider;
+use phpmock\Mock;
+use phpmock\MockBuilder;
+use PHPUnit\Framework\TestCase;
+
+class Test_Jetpack_JITM extends TestCase {
+	public function setUp() {
+		$this->mock_add_action();
+		$this->mock_do_action();
+		$this->mock_wp_enqueue_script();
+
+		// input/output of these functions doesn't matter right now, they just need to exist
+		$this->mock_empty_function( 'wp_register_style' );
+		$this->mock_empty_function( 'plugins_url' );
+		$this->mock_empty_function( 'wp_style_add_data' );
+		$this->mock_empty_function( 'wp_enqueue_style' );
+		$this->mock_empty_function( 'wp_localize_script' );
+		$this->mock_empty_function( 'esc_url_raw' );
+		$this->mock_empty_function( 'rest_url' );
+		$this->mock_empty_function( 'esc_html__' );
+	}
+
+	public function tearDown() {
+		Mock::disableAll();
+		\Mockery::close();
+		$this->clear_added_actions();
+		$this->clear_enqueued_scripts();
+	}
+
+	public function test_jitm_disabled_by_filter() {
+		$this->mock_filters( array(
+			array( 'jetpack_just_in_time_msgs', false, false ),
+		) );
+
+		$jitm = new JITM();
+		$this->assertFalse( $jitm->register() );
+
+		$this->clear_mock_filters();
+	}
+
+	public function test_jitm_enabled_by_default() {
+		$this->mock_filters( array(
+			array( 'jetpack_just_in_time_msgs', false, true ),
+		) );
+
+		$jitm = new JITM();
+		$this->assertTrue( $jitm->register() );
+
+		$this->clear_mock_filters();
+	}
+
+	/**
+	 * This is an example of a test which uses Mockery to tests a class static method.
+	 *
+	 * It requires the runInSeparateProcess tag so that the class isn't already autoloaded.
+	 *
+	 * @runInSeparateProcess
```

Comment:
I'm concerned about using this as a common way to mock static methods. Judging by how much we use statics everywhere, we'll end up a bunch of `@runInSeparateProcess`/`runTestsInSeparateProcesses` tests, and this will end up making our tests **very slow**.

Just consider this comparison: I'm running the same 25 tests locally, see the difference if we run them in separate process vs the same process:

**Same process**
![](https://cldup.com/OzmwjKL2lC.png)

**Separate processes**
![](https://cldup.com/qUm9vtsvVq.png)

Now if we imagine this scaling up to a thousand tests - it would contribute to our tests between 10 and 20 times slower, which I think is a no-go. 

We've been discussing with @lezama about what alternatives we have, and he mentioned something about using traits to leverage easier mocking of static methods, but from what I noticed, it'd require some boilerplate in each class that uses static methods, so that's not great - ideally we'd want to be able to test things without too much boilerplate.

Another alternative I was thinking about was hacking around the autoloader and maybe setting up one mock per class in a base TestCase class, and actually use those mocks for mocking and stubbing static methods. This way we won't have to run each test in a separate process. Since regardless we'll end up with base test class, it would make sense to try that.

Reply:
> Another alternative I was thinking about was hacking around the autoloader and maybe setting up one mock per class in a base TestCase class

@tyxla and I are going to explore this maybe Today 👍

Reply:
I don't think we'll end up with nearly as many tests in separate processes as you think. First of all, most uses of static methods will be factored into instance methods. Secondly, the methods that remain will be mostly stateless without side effects (eg utility functions) so won't need mocking in most cases.